toad.8
.\"
.\" Copyright (c) 2021 jpz4085
.\" Copyright (c) 2016 Antoine Jacoutot <ajacoutot@openbsd.org>
.\" Copyright (c) 2013 M:tier Ltd.
.\"
.\" Permission to use, copy, modify, and distribute this software for any
.\" purpose with or without fee is hereby granted, provided that the above
.\" copyright notice and this permission notice appear in all copies.
.\"
.\" THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
.\" WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
.\" MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
.\" ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
.\" WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
.\" ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
.\" OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
.\"
.\"
.Dd $Mdocdate: March 25 2021 $
.Dt TOAD 8
.Os
.Sh NAME
.Nm toad
.Nd
device automounter for
.Xr hotplugd 8
.Sh SYNOPSIS
.Nm
.Bk -words
.Ar action
.Ar devclass
.Ar devname
.Ek
.Sh DESCRIPTION
.Nm
(Toad Opens All Devices)
is a utility meant to be started from the
.Xr hotplugd 8
.Em attach
and
.Em detach
scripts.
It will attempt to mount all supported partitions on the device under
.Pa /media/${USER}/device .
Where ${USER} is the active user
.Em login
name and
.Em device
is either the partition volume label or the disk label description
followed by a partition number (ex. Sandisk Cruzer-p1). Mounting
uses the hotplug-diskmount(8) command and follows the
.Em udev
hierarchy in Linux which allows interaction with GLib/GIO's GUnixMount.
.Pp
Users outside the default UID range (1000...60000) are ignored (nothing will
be done).
.Pp
The arguments are as follows:
.Bl -tag -width Ds
.It Ar action
Can be
.Ar attach
(for device attachment)
or
.Ar detach
(for device detachment).
.It Ar devclass
Integer corresponding to the device class as described in the
.Aq Pa sys/device.h
header file,  e.g.\&
.Pa 2
for disk drives.
.It Ar devname
Device name as listed in
.Xr MAKEDEV 8 ,
and the unit number, e.g.\&
.Pa sd1 .
.El
.Pp
Detection of the currently active user is done using ConsoleKit and DBus,
.Nm
will not do anything unless these are properly setup and running.
Obviously,
.Xr hotplugd 8
must be running as well.
.Xr rcctl 8
can be used to enable and start
.Xr hotplugd 8
and
.Xr dbus-daemon 1 :
.Bd -literal -offset indent
.Ev rcctl enable hotplugd messagebus
.Ev rcctl start hotplugd messagebus
.Ed
.Sh EXAMPLES
This sample code can be used for both the
.Em attach
and
.Em detach
scripts:
.Bd -literal -offset indent
#!/bin/sh

DEVCLASS=$1
DEVNAME=$2

case $DEVCLASS in
2)
	/usr/local/libexec/toad ${0##*/} $DEVCLASS $DEVNAME
	;;
esac
.Ed
.Sh SEE ALSO
.Xr dbus-daemon 1 ,
.Xr hotplugd 8 ,
.Xr hotplug-diskmount 8
.Sh AUTHORS
.An Antoine Jacoutot Aq Mt antoine@mtier.org


ChangeLog
v2.0 - Wed Mar 31 18:05:41 EDT 2021
------------------------------------
- perform mount operations using enhanced hotplug-diskmount(8)
- works with fuse FS drivers supported by hotplug-diskmount(8)
- removed old mount point management code which is not required
- fixed format of dbus session file name needed for gdbus_call
- run makewhatis after install to update manual index database
- add uninstall function to remove files and reindex database

v1.14 - UNRELEASED
------------------------------------

v1.13 - Mon Nov 30 20:48:33 CET 2020
------------------------------------
- skip device on errors
- properly skip RAID devices
- skip already mounted partitions

v1.12 - Mon Nov 30 20:48:33 CET 2020
------------------------------------
- remove some debug output
- redirect gdbus calls stdout to /dev/null

v1.11 - Sun Nov 29 11:03:54 CET 2020
------------------------------------
- re-introduce the gdbus calls now that we have a way to connect to the user
  session; this will automatically open your file manager on plugging and
  display notifications on the Desktop

v1.10 - Sun Nov 29 10:59:52 CET 2020
------------------------------------
- no changes

v1.9 - Sat Nov 28 17:43:27 CET 2020
-----------------------------------
- stop playing games with gdbus; it's useless as we need XAUTHORITY of the user
  which we cannot get after forking

v1.8 - Sun Feb 10 15:03:09 CET 2019
-----------------------------------
- only works with regular users (uid 1000...60000)
- add support for ejecting cd-rom
- remove support for firmware installation, it can easily be done using
  fw_update(1) in the hotplugd(8) attach script

v1.7 - Wed Jan 23 04:37:24 CET 2019
-----------------------------------
- big rewrite to use pkexec(1), needs glib2 >=2.58.3p0 (OpenBSD package)
- add net device to the attach script sample

v1.6 - Thu Feb 25 13:53:42 CET 2016
-----------------------------------
- only react on explicit hotplug(4) events

v1.5 - Thu Feb 25 13:28:36 CET 2016
-----------------------------------
- add support for fw_update(1); from Robert Nagy (robert@mtier.org)

v1.4 - Wed Dec 23 15:09:54 CET 2015
-----------------------------------
- drop support for opening the default file manager
- fix output when running from the cli

v1.3 - Mon Jul 20 02:49:04 CET 2015
-----------------------------------
- display an error message if we cannot mount a device
- properly detect our xmessage compatible command

v1.2 - Thu Apr 23 15:49:36 CET 2015
-----------------------------------
- rc.d(8) script for dbus-daemon(1) was renamed (Adam Wolk)
- tweak error messages
- fix typo

v1.1 - Sun Mar  2 17:24:40 CET 2014
-----------------------------------
- make toad(8) more resilient to weird corner cases

v1.0 - Sun Mar  2 13:48:27 CET 2014
-----------------------------------
- drop all secondary groups inherited from root when we fork and setgid
- mount point and device ownership are now created and set by the child
- add a call to restore_dev in broom_sweep
- do not try to mount 'c': not our job to deal with broken disklabels
- make toad(8) more resilient, sanitize all return codes
- extend documentation a bit after users'feedback
- do nothing if uid is lower than 1000 (OpenBSD default UID_MIN)
- display is not optional since we have no ConsoleKit PAM support on OpenBSD

v0.9 - Fri Dec 27 10:05:28 CET 2013
-----------------------------------
- Add UDF support
- Add a sample attach+detach script
- Properly check for devices locked in read-only mode
- Fix zenity window height and width

v0.8 - Wed Dec 11 12:54:22 CET 2013
-----------------------------------
- Ignore softraid(4) crypto disk attachement
- Only change the device owner, not the group
- Fix ejecting optical media, gio passes the block device path as
  argument to eject(1), not the raw one

v0.7 - Mon Dec  9 12:32:09 CET 2013
-----------------------------------
- toadd: fix path to toad(8)

v0.6 - Mon Dec  9 10:08:37 CET 2013
-----------------------------------
- Exit and clean when no partition is found on the device
- Only attempt to mount the 'c' partition if no other partition is
  found and the label is MSDOS
- Try to mount R/O if the initial mount command failed (and warn the user)

v0.5 - Thu Dec  5 15:19:24 CET 2013
-----------------------------------
- Fix compilation with WARNINGS set to "Yes"
  Giovanni Bechis (giovanni@openbsd.org)

v0.4 - Wed Dec  4 14:35:27 CET 2013
-----------------------------------
- do not hardcode path to toad(8)
- stricter permissions on mount path

v0.3 - Tue Oct  8 18:43:31 CET 2013
-----------------------------------
- fix install target
- minor cleanups

v0.2 - Thu Oct  4 14:18:58 UTC 2013
-----------------------------------
- toadd: new option `-w' to specify the optical disk polling interval
- toadd: new option `-d' to run in foreground with debug output

v0.1 - Thu Oct  3 14:18:58 UTC 2013
-----------------------------------
- Initial import of toad-0.1


toad.pl

#!/usr/bin/perl
#
# Copyright (c) 2021 jpz4085
# Copyright (c) 2016, 2019, 2020 Antoine Jacoutot <ajacoutot@openbsd.org>
# Copyright (c) 2013, 2014, 2015, 2016 M:tier Ltd.
#
# Permission to use, copy, modify, and distribute this software for any
# purpose with or without fee is hereby granted, provided that the above
# copyright notice and this permission notice appear in all copies.
#
# THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
# WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
# MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
# ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
# WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
# ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
# OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
#
# Author: Antoine Jacoutot <antoine@mtier.org>

use strict;
use warnings;

use File::Path qw(make_path);
use Net::DBus;
use User::pwent qw(:FIELDS);

if ( $< != 0 ) {
	print "need root privileges\n";
	exit (1);
}

sub usage {
	print "usage: $0 attach|detach devclass device\n";
	exit (1);
}

if (@ARGV < 3) { usage (); }

my ($action, $devclass, $devname) = @ARGV;
my ($login, $uid, $gid, $display, $home) = get_active_user_info ();
my $dbus_session_bus_address = get_dbus_session_bus_address ();
my $mounttop = "/media/$login";
my $mountbase;
my $mountfuse;
my $devtype;
my $devmax;
my $pkrulebase = "/etc/polkit-1/rules.d/45-toad-$login";

sub broom_sweep {
	unlink glob "$pkrulebase-$devname?.rules";

	if ($devtype eq 'cd') {
		exec("/usr/local/libexec/hotplug-diskmount -d $mounttop cleanup $devname");
	}
}

sub create_pkrule {
	my($devname, $part, $mountbase) = @_;
	my $pkrule = "$pkrulebase-$devname$part.rules";

	unless(open PKRULE, '>'.$pkrule) {
		die "Unable to create $pkrule\n";
	}

	print PKRULE "polkit.addRule(function(action, subject) {\n";
	print PKRULE "  if (action.id == \"org.freedesktop.policykit.exec\" &&\n";
	print PKRULE "    action.lookup(\"program\") == \"/sbin/umount\" &&\n";
	print PKRULE "    action.lookup(\"command_line\") == \"/sbin/umount $mountbase\") {\n";
	print PKRULE "    if (subject.local && subject.active && subject.user == \"$login\") {\n";
	print PKRULE "      return polkit.Result.YES;\n";
	print PKRULE "    }\n";
	print PKRULE "  }\n";
	print PKRULE "});\n";

	if ($devtype eq 'cd') {
		print PKRULE "polkit.addRule(function(action, subject) {\n";
		print PKRULE "  if (action.id == \"org.freedesktop.policykit.exec\" &&\n";
		print PKRULE "    action.lookup(\"program\") == \"/bin/eject\" &&\n";
		print PKRULE "    action.lookup(\"command_line\") == \"/bin/eject /dev/$devname$part\") {\n";
		print PKRULE "    if (subject.local && subject.active && subject.user == \"$login\") {\n";
		print PKRULE "      return polkit.Result.YES;\n";
		print PKRULE "    }\n";
		print PKRULE "  }\n";
		print PKRULE "});\n";
	}

	close PKRULE;
}

sub gdbus_call {
	my($action, $args) = @_;
	my $cmd = "gdbus call -e";

	my $pid = fork ();
	if (!defined ($pid)) {
		die "could not fork: $!";
	} elsif ($pid) {
		if (waitpid ($pid, 0) > 0) {
			if ($? >> 8 ne 0) {
				return (1);
			}
		}
	} else {
		$( = $) = "$gid $gid";
		$< = $> = $uid;

		$ENV{"DISPLAY"} = $display;
		$ENV{"HOME"} = $home;
		$ENV{"DBUS_SESSION_BUS_ADDRESS"} = $dbus_session_bus_address;

		if ($action eq 'notify') {
			print ("$args\n");
			if (defined ($dbus_session_bus_address)) {
				$cmd .= " -d org.freedesktop.Notifications";
				$cmd .= " -o /org/freedesktop/Notifications";
				$cmd .= " -m org.freedesktop.Notifications.Notify";
				$cmd .= " toad 42 drive-harddisk-usb";
				$cmd .= " \"Toad\" \"$args\" [] {} 5000 >/dev/null";
				system($cmd);
			}
		} elsif ($action eq 'open-fm') {
			if (defined ($dbus_session_bus_address)) {
				$cmd .= " -d org.freedesktop.FileManager1";
				$cmd .= " -o /org/freedesktop/FileManager1";
				$cmd .= " -m org.freedesktop.FileManager1.ShowFolders";
				$cmd .= " '[\"file://$args\"]' \"\" >/dev/null";
				system($cmd);
			}
		}
		# exit the child
		exit (0);
	}
}

sub get_active_user_info {
	my $system_bus = Net::DBus->system;
	my $ck_service = $system_bus->get_service ('org.freedesktop.ConsoleKit');
	my $ck_manager = $ck_service->get_object ('/org/freedesktop/ConsoleKit/Manager');

	for my $session_id (@{$ck_manager->GetSessions ()}) {
		my $ck_session = $ck_service->get_object ($session_id);
		next unless $ck_session->IsActive ();

		my $uid = $ck_session->GetUnixUser ();
		getpwuid ($uid) || die "no $uid user: $!";
		next unless ($uid >= 1000 && $uid <= 60000);

		my $display = $ck_session->GetX11Display ();
		next unless length ($display);

		my $gid = $pw_gid;
		my $login = $pw_name;
		my $home = $pw_dir;

		return ($login, $uid, $gid, $display, $home);
	}
}

sub get_dbus_session_file {
	my $id;
	my $machine_id = "/etc/machine-id";

	if (open my $fh, "<", $machine_id) {
		read $fh, $id, -s $fh;
		close $fh;
	} else {
		print "Can't open file \"$machine_id\"\n";
		return;
	}

	$id =~ s/\R//g; # drop line break
	$display =~ s/^.{3}//;

	return "$home/.dbus/session-bus/$id-$display";
}

sub get_dbus_session_bus_address {
	my $dbus_session_file = get_dbus_session_file ();

	if (!defined ($dbus_session_file)) {
		return;
	}

	if (open my $fh, "<", $dbus_session_file) {
		while (<$fh>) {
			chomp;
			my ($l, $r) = split /=/, $_, 2;
			if ($l eq "DBUS_SESSION_BUS_ADDRESS") {
				$r =~ s/'//g;
				return ($r);
			}
		}
		close $fh;
	} else {
		print "Can't open file \"$dbus_session_file\"\n";
	}
}

sub get_mount_point {
	my($devname, $part) = @_;
	my $mntgrep;
	my $mntpath;
	my $i = 3;

	do {
		$mntgrep = `/sbin/mount | /usr/bin/grep $devname$part | /usr/bin/awk \'{print \$$i}\'`;
		chomp ($mntgrep);
		if ($mntgrep ne 'type') {
			if ($i >= 4) {
				$mntpath .= "\\ $mntgrep";
			} else {
				$mntpath = $mntgrep;
			}
		}
		$i++;
	} while ($mntgrep ne 'type');

	return ($mntpath);
}

sub get_ntfs_label {
	my($devname, $ntfspart) = @_;
	my $mntntfs;
	my $mntexfat;

	$mntntfs = `/usr/local/sbin/ntfslabel /dev/$devname$ntfspart 2>&1`;
	$mntexfat = `/usr/local/sbin/exfatlabel /dev/$devname$ntfspart 2>&1`;
	chomp ($mntntfs, $mntexfat);
	if (index($mntntfs, 'NTFS signature is missing.') != -1) {
		if ($mntexfat eq '') {
			$mntexfat = "NONAME";
		}
		$mntexfat =~ s/ /\\ /g;
		return ("$mounttop/$mntexfat");
	} else {
		if ($mntntfs eq '') {
			$mntntfs = "NONAME";
		}
		$mntntfs =~ s/ /\\ /g;
		return ("$mounttop/$mntntfs");
	}		
}

sub get_parts {
	my @allparts;
	my @ntfsparts;
	my @supportedfs = ('MSDOS', 'NTFS', '4.2BSD', 'ext2fs', 'ISO9660', 'UDF');

	foreach my $fs (@supportedfs) {
		my $fsmatch = `/sbin/disklabel $devname 2>/dev/null | /usr/bin/grep " $fs "`;
		while ($fsmatch =~ /([^\n]+)\n?/g) {
			my @part = split /:/, $1;
			$part[0] =~ s/ //g;
			push (@allparts, $part[0]);
			if ($fs eq 'NTFS') {
				push (@ntfsparts, $part[0]);
			}
		}
	}

	return (\@allparts, \@ntfsparts);
}

sub mount_device {
	my @parts;
	my @ntfsp;

	# XXX skip device on error (e.g. DIOCGDINFO) or softraid(4) attachment
	if (system ("set -o pipefail; /sbin/disklabel $devname 2>/dev/null | ! grep -qw RAID") != 0) {
		return (0);
	}

	if ($devtype eq 'cd') {
		@parts = 'a';
	} else {
		my ($allparts, $ntfsparts) = get_parts ();
		foreach my $part (@$allparts) {
			if ($part !~ 'c$') {
				push @parts, $part;
			}
		}
		foreach my $ntfs (@$ntfsparts) {
			if ($ntfs !~ 'c$') {
				push @ntfsp, $ntfs;
			}
		}
	}

	unless (@parts) {
		gdbus_call ("notify", "No supported partition found on device $devname");
		return (0);
	}

	if (@ntfsp) {
		foreach my $ntfs (@ntfsp) {
			$ntfs =~ s/^\s+//;
			$mountfuse = get_ntfs_label ($devname, $ntfs);
			create_pkrule ($devname, $ntfs, "$mountfuse");
		}
	}

	if (!-d $mounttop) {
		my $mountrw = system("/usr/local/libexec/hotplug-diskmount -d $mounttop init 2>&1");
		if ($mountrw != 0) {
			gdbus_call ("notify", "Cannot create mount folder $mounttop");
			exit (1);
		}
		sleep (1);
	}
	my $mountrw = system("/usr/local/libexec/hotplug-diskmount -d $mounttop attach -m 0700 -u $login $devname 2>&1");
	if ($mountrw != 0) {
		gdbus_call ("notify", "Cannot mount partitions on device $devname");
		exit (1);
	}
	my $pcount = @parts;
	sleep ($pcount);

	foreach my $part (@parts) {
		unless (grep $_ eq $part, @ntfsp) {
			$part =~ s/^\s+//;
			$mountbase = get_mount_point ($devname, $part);
			create_pkrule ($devname, $part, "$mountbase");
			gdbus_call ("open-fm", "$mountbase");
		} else {
			gdbus_call ("open-fm", "$mountfuse");
		}
	}
}

if ($devclass == 2) {
	$devtype = 'usb';
	$devmax = 10;
} elsif ($devclass == 9) {
	$devtype = 'cd';
	$devmax = 2;
} else {
	gdbus_call ("notify", "Device type not supported");
	exit (1);
}

if ($action eq 'attach') {
	if (!defined ($login) || !defined ($uid) || !defined ($gid)) {
		print "ConsoleKit: user does not own the active session\n";
		exit (1);
	}
	if ($devtype eq 'cd' || $devtype eq 'usb') { mount_device (); }
} elsif ($action eq 'detach') {
	if ($devtype eq 'cd' || $devtype eq 'usb') { broom_sweep (); }
} else { usage (); }
